src/test/mon/test_election.cc

```diff
@@ -0,0 +1,369 @@
+    void trigger_new_election() { logic.start(); }
+    int get_my_rank() { return rank; }
+    void propose_to_peers(epoch_t e) {
+      for (int i = 0; i < parent->get_paxos_size(); ++i) {
+	if (i == rank) continue;
+	parent->propose_to(rank, i, e);
+      }
+    }
+    void reset_election() {
+      _start();
+      logic.start();
+    }
+    bool ever_participated() { return ever_joined; }
+    unsigned paxos_size() { return parent->get_paxos_size(); }
+    void cancel_timer() {
+      timer_steps = -1;
+    }
+    void reset_timer(int steps) {
+      cancel_timer();
+      timer_steps = 3+steps; // FIXME? magic number, current step + roundtrip
+      timer_election = true;
+    }
+    void start_victory_timer() {
+      cancel_timer();
+      timer_election = false;
+      timer_steps = 3; // FIXME? current step + roundtrip
+    }
+    void _start() {
+      reset_timer(0);
+      quorum.clear();
+    }
+    void _defer_to(int who) {
+      parent->defer_to(rank, who, logic.get_epoch());
+      reset_timer(0);
+    }
+    void message_victory(const set<int>& members) {
+      for (auto i : members) {
+	if (i == rank) continue;
+	parent->claim_victory(rank, i, logic.get_epoch(), members);
+      }
+      start_victory_timer();
+      quorum = members;
+      victory_accepters = 1;
+    }
+    bool is_current_member(int rank) { return quorum.count(rank) != 0; }
+    void receive_propose(int from, epoch_t e) {
+      logic.receive_propose(from, e);
+    }
+    void receive_ack(int from, epoch_t e) {
+      if (e < logic.get_epoch())
+	return;
+      logic.receive_ack(from, e);
+    }
+    void receive_victory_claim(int from, epoch_t e, const set<int>& members) {
+      if (e < logic.get_epoch())
+	return;
+      if (logic.receive_victory_claim(from, e)) {
+	quorum = members;
+	cancel_timer();
+	parent->accept_victory(rank, from, e);
+      }
+    }
+    void receive_victory_ack(int from, epoch_t e) {
+      if (e < logic.get_epoch())
+	return;
+      ++victory_accepters;
+      if (victory_accepters == static_cast<int>(quorum.size())) {
+	cancel_timer();
+	for (int i : quorum) {
+	  parent->electors[i]->ever_joined = true;
+	}
+      }
+    }
+    void election_timeout() {
+      cerr << "election epoch " << logic.get_epoch()
+	   << " timed out for " << rank
+	   << ", electing me:" << logic.electing_me
+	   << ", acked_me:" << logic.acked_me << std::endl;
+      logic.end_election_period();
+    }
+    void victory_timeout() {
+      cerr << "victory epoch " << logic.get_epoch()
+	   << " timed out for " << rank
+	   << ", electing me:" << logic.electing_me
+	   << ", acked_me:" << logic.acked_me << std::endl;
+      reset_election();
+    }
+    void notify_timestep() {
+      assert(timer_steps != 0);
+      if (timer_steps > 0) {
+	--timer_steps;
+      }
+      if (timer_steps == 0) {
+	if (timer_election) {
+	  election_timeout();
+	} else {
+	  victory_timeout();
+	}
+      }
+    }
+  };
+}
+
+Election::Election(int c) : count(c)
```

Comment:
why put the declaration in the anonymous namespace. while put its implementation out of it?

Reply:
Uh, I think I cargo-culted the namespace from another test. No idea why.
As it seems to work without that I've just dropped it entirely.